src/pybind/mgr/volumes/fs/clone_stats.py

```diff
@@ -0,0 +1,215 @@
+                                ci.dst_group_name, ci.dst_subvol_name,
+                                SubvolumeOpType.CLONE_INTERNAL) \
+                                as (_, _, dst_subvol):
+            ci.dst_path = dst_subvol.path
+            log.debug(f'destination subvolume path for clone - {ci.dst_path}')
+
+        log.debug('finished collecting info for cloning destination')
+
+    def _get_clone_src_info(self, fs_handle, ci):
+        log.debug('collecting info for cloning source')
+
+        with open_clone_subvol_pair_in_vol(self.volclient, self.volspec,
+                                           ci.volname, ci.dst_group_name,
+                                           ci.dst_subvol_name) as \
+                                           (dst_subvol, src_subvol, snap_name):
+            ci.src_group_name = src_subvol.group_name
+            ci.src_subvol_name = src_subvol.subvolname
+            ci.src_path = src_subvol.snapshot_data_path(snap_name)
+            log.debug(f'source subvolume path for clone - {ci.src_path}')
+
+        log.debug('finished collecting info for cloning source')
+
+    def _get_info_for_all_clones(self):
+        clones:list[CloneInfo] = []
+
+        log.debug('collecting all entries in clone index...')
+        volnames = list_volumes(self.volclient.mgr)
+        for volname in volnames:
+            with open_volume_lockless(self.volclient, volname) as fs_handle:
+                with open_clone_index(fs_handle, self.volspec) as clone_index:
+                    clone_index_path = clone_index.path
+                    # get clone in order in which they were launched, this
+                    # should be same as the ctime on clone entry.
+                    clone_index_entries = clone_index.list_entries_by_ctime_order()
+                    log.debug('finished collecting all clone index entries, '
+                              f'found {len(clones)} clone index entries')
+
+                log.debug('collecting info for clones found through clone index '
+                         'entries...')
+                for ce in clone_index_entries:
+                    ci = CloneInfo(volname)
+
+                    try:
+                        self._get_clone_dst_info(fs_handle, ci, ce,
+                                                 clone_index_path)
+                        self._get_clone_src_info(fs_handle, ci)
+                    except ObjectNotFound as e:
+                        log.info('Exception ObjectNotFound was raised. Apparently '
+                                 'entry in clone index was removed because one of '
+                                 'the clone job(s) has completed/cancelled, '
+                                 'therefore ignoring and proceeding. '
+                                 f'Printing the exception: {e}')
+                        continue
+                    except VolumeException as e:
+                        if e.error_str != 'error fetching subvolume metadata':
```

Comment:
This seems prone to false negatives and is a no-no.

Reply:
We can catch EINVAL instead but that won't tell us if EINVAL was specifically raised for "error fetching subvolume metadata". So checking error message is pretty important. Had a chat with Venky about this just now, he has the opinion that this should be avoided unless where it's really needed. And he intends to revisit this comment to check that.

Reply:
I had a look at this and the conclusion is that error string checks are to be avoided. So, please let's not do this. If you need to handle special cases, then return district error codes to check against in the caller.

Reply:
Separate error code for fetching metdata? I can't tell what's error code is appropriate besides EINVAL. Please recommend one.

Also, this would cause change in every other location where we catch the error. Will that be fine?

Reply:
> Separate error code for fetching metdata? I can't tell what's is appropriate besides EINVAL. Please recommend one.

This is catching exceptions raised by _get_clone_dst_info() and _get_clone_src_info() which will end up calling subvolume.py:path() or subvolume.py:_get_clone_source() where MetadataMgrException raises -EINVAL, "...". So, I still think you can just handle EINVAL case here. If you are worried about EINVAL being returned from other call in the call stack (which may not be a metadata fetch problem) which might get silently ignored, we need to understand which case is that and for those cases separate errno's can be used . So, unless there is a list of calls that could return EINVAL that might get masked, I would suggest we work with comparing errno's instead.

Reply:
This PR intends to move this piece of code from `stats_util.py` to `clone_stats.py` but improving it on this PR is not appropriate. I've opened a separate PR to make the suggested improvement in mgr/vol code: https://github.com/ceph/ceph/pull/61004.